paasta_tools/utils.py

```diff
@@ -2854,13 +2854,33 @@ def to_bytes(obj: Any) -> bytes:
         return str(obj).encode('UTF-8')
 
 
+TLS = threading.local()
+
+
+@contextlib.contextmanager
+def set_paasta_print_file(file: Any) -> Iterator[None]:
+    TLS.paasta_print_file = file
+    yield
+    TLS.paasta_print_file = None
+
+
 def paasta_print(*args: Any, **kwargs: Any) -> None:
     f = kwargs.pop('file', sys.stdout)
-    f = getattr(f, 'buffer', f)
-    end = to_bytes(kwargs.pop('end', '\n'))
-    sep = to_bytes(kwargs.pop('sep', ' '))
+    f = getattr(TLS, 'paasta_print_file', f)
+    buf = getattr(f, 'buffer', None)
+    # Here we're assuming that the file object works with strings and its
+    # `buffer` works with bytes. So, if the file object doesn't have `buffer`,
+    # we output via the file object itself using strings.
+    obj_to_arg: Callable[[Any], Any]
+    if buf is not None:
+        f = buf
+        obj_to_arg = to_bytes
+    else:
+        def obj_to_arg(o: Any) -> str: return to_bytes(o).decode('UTF-8', errors="ignore")
```

Comment:
the `errors="ignore"` seems dangerous - can you explain?

Reply:
`to_bytes()` converts its string and object args to UTF-8-encoded byte sequences.  At the same time, it doesn't convert byte sequence args at all, so there is no guaranty that a byte sequences returned from `to_bytes()` can be safely converted back to a string using UTF-8 encoding. So, there is a possibility that we can have `UnicodeEncodeError` exception if we try to decode `to_bytes()`'s return value to a string using UTF-8 encoding.

At the same time, `print_paasta()` was introduced partially to prevent `UnicodeEncodeError: 'ascii' codec can't encode character` errors while printing.  So, to satisfy that, we need to make sure, that decoding `to_bytes()`'s return value is always successful.  To do that, I've added `error="ignore"` flag.